File: index_page_manager.cpp
#include "index_page_manager.h"

#include <glog/logging.h>

#include <cassert>
#include <cstdint>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "async_io_manager.h"
#include "error.h"
#include "kv_options.h"
#include "mem_index_page.h"
#include "page_mapper.h"
#include "replayer.h"
#include "root_meta.h"
#include "task.h"
#include "types.h"

namespace eloqstore
{
IndexPageManager::IndexPageManager(AsyncIoManager *io_manager)
    : io_manager_(io_manager)
{
    active_head_.EnqueNext(&active_tail_);
}

IndexPageManager::~IndexPageManager()
{
    for (auto &[tbl_id, meta] : tbl_roots_)
    {
        meta.mapper_ = nullptr;
    }
}

const Comparator *IndexPageManager::GetComparator() const
{
    return io_manager_->options_->comparator_;
}

MemIndexPage *IndexPageManager::AllocIndexPage()
{
    MemIndexPage *next_free = free_head_.DequeNext();

    while (next_free == nullptr)
    {
        if (!IsFull())
        {
            auto &new_page =
                index_pages_.emplace_back(std::make_unique<MemIndexPage>());
            next_free = new_page.get();
        }
        else
        {
            bool success = Evict();
            if (!success)
            {
                // There is no page to evict because all pages are pinned.
                // Tasks trying to allocate new pages should rollback to unpin
                // pages in the task's traversal stack.
                return nullptr;
            }
            next_free = free_head_.DequeNext();
        }
    }
    assert(next_free->IsDetached());
    assert(!next_free->IsPinned());
    return next_free;
}

void IndexPageManager::FreeIndexPage(MemIndexPage *page)
{
    assert(page->IsDetached());
    assert(!page->IsPinned());
    free_head_.EnqueNext(page);
}

void IndexPageManager::EnqueueIndexPage(MemIndexPage *page)
{
    if (page->prev_ != nullptr)
    {
        assert(page->next_ != nullptr);
        page->Deque();
    }
    assert(page->prev_ == nullptr && page->next_ == nullptr);
    active_head_.EnqueNext(page);
}

bool IndexPageManager::IsFull() const
{
    // Calculate current total memory usage
    size_t current_size = index_pages_.size() * Options()->data_page_size;
    return current_size >= Options()->index_buffer_pool_size;
}

std::pair<RootMeta *, KvError> IndexPageManager::FindRoot(
    const TableIdent &tbl_id)
{
    auto load_meta = [this](const TableIdent &tbl_id, RootMeta *meta)
    {
        // load manifest file
        auto [manifest, err] = IoMgr()->GetManifest(tbl_id);
        CHECK_KV_ERR(err);

        // replay
        Replayer replayer(Options());
        err = replayer.Replay(manifest.get());
        if (err != KvError::NoError)
        {
            LOG(ERROR) << "load evicted table: replay failed";
            return err;
        }

        meta->root_id_ = replayer.root_;
        meta->ttl_root_id_ = replayer.ttl_root_;
        auto mapper = replayer.GetMapper(this, &tbl_id);
        MappingSnapshot *mapping = mapper->GetMapping();
        meta->mapper_ = std::move(mapper);
        meta->mapping_snapshots_.insert(mapping);
        meta->Pin();
        meta->manifest_size_ = replayer.file_size_;
        meta->next_expire_ts_ = 0;
        meta->compression_->LoadDictionary(std::move(replayer.dict_bytes_));
        if (meta->ttl_root_id_ != MaxPageId)
        {
            // For simplicity, we initialize next_expire_ts_ to 1,
            // ensuring the next write operation will trigger a TTL check.
            meta->next_expire_ts_ = 1;
        }
        return KvError::NoError;
    };

    while (true)
    {
        auto [it, inserted] = tbl_roots_.try_emplace(tbl_id);
        RootMeta *meta = &it->second;

        if (inserted)
        {
            // Try to load metadata from persistent storage.
            meta->locked_ = true;
            KvError err = load_meta(it->first, meta);
            meta->waiting_.WakeAll();
            if (err != KvError::NoError)
            {
                tbl_roots_.erase(tbl_id);
                return {nullptr, err};
            }
            meta->locked_ = false;
        }
        else if (meta->locked_)
        {
            // Blocked by other loading/evicting operation.
            meta->waiting_.Wait(ThdTask());
            continue;
        }

        if (meta->mapper_ == nullptr)
        {
            // Partition not found. Possible causes:
            // 1. During the initial write to a non-existent partition,
            //    WriteTask creates a stub RootMeta (with mapper=nullptr).
            // 2. A MemIndexPage referencing this stub RootMeta is created.
            // 3. WriteTask aborts, but the stub RootMeta cannot be cleared
            //    because it is still referenced by the MemIndexPage.
            return {meta, KvError::NotFound};
        }
        return {meta, KvError::NoError};
    }
}

KvError IndexPageManager::MakeCowRoot(const TableIdent &tbl_ident,
                                      CowRootMeta &cow_meta)
{
    auto [meta, err] = FindRoot(tbl_ident);
    if (err == KvError::NoError)
    {
        // Makes a copy of the mapper.
        auto new_mapper = std::make_unique<PageMapper>(*meta->mapper_);
        cow_meta.root_id_ = meta->root_id_;
        cow_meta.ttl_root_id_ = meta->ttl_root_id_;
        cow_meta.mapper_ = std::move(new_mapper);
        cow_meta.old_mapping_ = meta->mapper_->GetMappingSnapshot();
        cow_meta.manifest_size_ = meta->manifest_size_;
        cow_meta.next_expire_ts_ = meta->next_expire_ts_;
        if (meta->compression_->Dirty())
        {
            // This only happens when the dictionary is built from values with
            // expired timestamps. If eloqstore stops before any new value
            // arrives, this dictionary can be discarded since no value has been
            // written. Otherwise, the dictionary can still be used to compress
            // subsequent values.
            assert(cow_meta.manifest_size_ == 0);
        }
        cow_meta.compression_ = meta->compression_;
    }
    else if (err == KvError::NotFound)
    {
        // It is the WriteTask's responsibility to clean up this stub RootMeta
        // if it aborted.
        auto [tbl_it, _] = tbl_roots_.try_emplace(tbl_ident);
        const TableIdent *tbl_id = &tbl_it->first;
        auto mapper = std::make_unique<PageMapper>(this, tbl_id);
        std::shared_ptr<MappingSnapshot> mapping = mapper->GetMappingSnapshot();
        cow_meta.root_id_ = MaxPageId;
        cow_meta.ttl_root_id_ = MaxPageId;
        cow_meta.mapper_ = std::move(mapper);
        cow_meta.old_mapping_ = std::move(mapping);
        cow_meta.manifest_size_ = 0;
        cow_meta.next_expire_ts_ = 0;
        cow_meta.compression_ =
            std::make_shared<compression::DictCompression>();
        meta = &tbl_it->second;
    }
    else
    {
        return err;
    }
    auto it = meta->mapping_snapshots_.insert(cow_meta.mapper_->GetMapping());
    CHECK(it.second);
    meta->Pin();  // Referenced by new MappingSnapshot.
    return KvError::NoError;
}

void IndexPageManager::UpdateRoot(const TableIdent &tbl_ident,
                                  CowRootMeta new_meta)
{
    auto tbl_it = tbl_roots_.find(tbl_ident);
    assert(tbl_it != tbl_roots_.end());
    RootMeta &meta = tbl_it->second;
    meta.root_id_ = new_meta.root_id_;
    meta.ttl_root_id_ = new_meta.ttl_root_id_;
    if (meta.mapper_ != nullptr && !Options()->data_append_mode)
    {
        assert(new_meta.mapper_ != nullptr);
        MappingSnapshot *prev_snapshot = meta.mapper_->GetMapping();
        prev_snapshot->next_snapshot_ = new_meta.mapper_->GetMappingSnapshot();
    }
    meta.mapper_ = std::move(new_meta.mapper_);
    meta.manifest_size_ = new_meta.manifest_size_;
    meta.next_expire_ts_ = new_meta.next_expire_ts_;
    meta.compression_ = std::move(new_meta.compression_);
}

std::pair<MemIndexPage *, KvError> IndexPageManager::FindPage(
    MappingSnapshot *mapping, PageId page_id)
{
    while (true)
    {
        // First checks swizzling pointers.
        MemIndexPage *idx_page = mapping->GetSwizzlingPointer(page_id);
        if (idx_page == nullptr)
        {
            // This is the first request to load the page.
            MemIndexPage *new_page = AllocIndexPage();
            if (new_page == nullptr)
            {
                return {nullptr, KvError::OutOfMem};
            }
            FilePageId file_page_id = mapping->ToFilePage(page_id);
            new_page->SetPageId(page_id);
            new_page->SetFilePageId(file_page_id);
            mapping->AddSwizzling(page_id, new_page);

            // Read the page async.
            auto [page, err] = IoMgr()->ReadPage(
                *mapping->tbl_ident_, file_page_id, std::move(new_page->page_));
            new_page->page_ = std::move(page);
            if (err != KvError::NoError)
            {
                new_page->waiting_.WakeAll();
                mapping->Unswizzling(new_page);
                FreeIndexPage(new_page);
                return {nullptr, err};
            }
            FinishIo(mapping, new_page);
            new_page->waiting_.WakeAll();
            return {new_page, KvError::NoError};
        }
        if (idx_page->IsDetached())
        {
            // This page is not loaded yet.
            idx_page->waiting_.Wait(ThdTask());
        }
        else
        {
            EnqueueIndexPage(idx_page);
            return {idx_page, KvError::NoError};
        }
    }
}

void IndexPageManager::FreeMappingSnapshot(MappingSnapshot *mapping)
{
    const TableIdent &tbl = *mapping->tbl_ident_;
    auto tbl_it = tbl_roots_.find(tbl);
    if (tbl_it == tbl_roots_.end())
    {
        return;
    }
    RootMeta &meta = tbl_it->second;
    // Puts back file pages freed in this mapping snapshot
    if (!mapping->to_free_file_pages_.empty())
    {
        assert(meta.mapper_ != nullptr);
        assert(!Options()->data_append_mode);
        auto pool =
            static_cast<PooledFilePages *>(meta.mapper_->FilePgAllocator());
        pool->Free(std::move(mapping->to_free_file_pages_));
    }
    auto n = meta.mapping_snapshots_.erase(mapping);
    CHECK(n == 1);
    meta.Unpin();
    EvictRootIfEmpty(tbl_it);
}

bool IndexPageManager::Evict()
{
    MemIndexPage *node = &active_tail_;

    do
    {
        while (node->prev_->IsPinned() && node->prev_ != &active_head_)
        {
            node = node->prev_;
        }

        // Has reached the head of the active list. Eviction failed.
        if (node->prev_ == &active_head_)
        {
            return false;
        }

        node = node->prev_;
        RecyclePage(node);
    } while (free_head_.next_ == nullptr);

    return true;
}

void IndexPageManager::EvictRootIfEmpty(const TableIdent &tbl_id)
{
    auto it = tbl_roots_.find(tbl_id);
    if (it != tbl_roots_.end())
    {
        EvictRootIfEmpty(it);
    }
}

void IndexPageManager::EvictRootIfEmpty(
    std::unordered_map<TableIdent, RootMeta>::iterator root_it)
{
    RootMeta &meta = root_it->second;

    if (meta.mapper_ == nullptr)
    {
        return;
    }
    CHECK(meta.ref_cnt_ != 0);

    if (meta.ref_cnt_ == 1)
    {
        if (meta.mapper_->UseCount() == 1)
        {
            // Check if mapping table is empty (no data pages)
            if (meta.mapper_->MappingCount() == 0 && meta.manifest_size_ > 0)
            {
                // Lock the rootmeta to prevent other FindRoot operations
                meta.locked_ = true;

                const TableIdent &tbl_id = root_it->first;

                // Clean up the table from io manager first

                // Note: it will also clean manifest when data_append = false
                // although it is not remove datafile
                // Manifest record max file page id, deleting it can lead to
                // creating existing files.
                IoMgr()->CleanManifest(tbl_id);

                // Wake up any waiting threads before erasing
                meta.waiting_.WakeAll();

                // Erase from memory - this will automatically destroy
                // meta.mapper_ and trigger MappingSnapshot destructor, but
                // FreeMappingSnapshot will return early since the table is
                // already erased
                tbl_roots_.erase(root_it);

                // Note: No need to unlock since we've erased the entry
                return;
            }

            // If mapping is not empty, we can directly erase the root since
            // ref_cnt == 1 means only the mapper itself holds the reference

            DLOG(INFO) << "metadata of " << root_it->first
                       << " is evicted (ref_cnt == 1)";
            tbl_roots_.erase(root_it);
        }
        else
        {
            // This is rare.
            DLOG(INFO) << "ref_cnt == 1 but mapper use count :"
                       << meta.mapper_->UseCount();
        }
    }
}

bool IndexPageManager::RecyclePage(MemIndexPage *page)
{
    assert(!page->IsPinned());
    auto tbl_it = tbl_roots_.find(*page->tbl_ident_);
    assert(tbl_it != tbl_roots_.end());
    RootMeta &meta = tbl_it->second;
    // Unswizzling the page pointer in all mapping snapshots.
    auto &mappings = meta.mapping_snapshots_;
    for (auto &mapping : mappings)
    {
        mapping->Unswizzling(page);
    }
    meta.Unpin();

    // Removes the page from the active list.
    page->Deque();
    EvictRootIfEmpty(tbl_it);
    assert(page->page_id_ != MaxPageId);
    assert(page->file_page_id_ != MaxFilePageId);
    page->page_id_ = MaxPageId;
    page->file_page_id_ = MaxFilePageId;
    page->tbl_ident_ = nullptr;

    FreeIndexPage(page);
    return true;
}

void IndexPageManager::FinishIo(MappingSnapshot *mapping,
                                MemIndexPage *idx_page)
{
    idx_page->tbl_ident_ = mapping->tbl_ident_;
    mapping->AddSwizzling(idx_page->GetPageId(), idx_page);

    if (idx_page->IsDetached())
    {
        auto tbl_it = tbl_roots_.find(*mapping->tbl_ident_);
        assert(tbl_it != tbl_roots_.end());
        tbl_it->second.Pin();
    }
    else
    {
        // index page is moved on physical position.
    }
    EnqueueIndexPage(idx_page);
}

KvError IndexPageManager::SeekIndex(MappingSnapshot *mapping,
                                    PageId page_id,
                                    std::string_view key,
                                    std::span<PageId> results,
                                    size_t &result_size_dst)
{
    auto [node, err] = FindPage(mapping, page_id);
    CHECK_KV_ERR(err);
    IndexPageIter idx_it{node, Options()};
    idx_it.Seek(key);
    PageId child_id = idx_it.GetPageId();
    if (!node->IsPointingToLeaf())
    {
        return SeekIndex(mapping, child_id, key, results, result_size_dst);
    }

    size_t result_size = 0;
    const size_t result_limit = results.size();
    if (child_id != MaxPageId && result_size < result_limit)
    {
        results[result_size++] = child_id;
    }

    while (result_size < result_limit && idx_it.HasNext() && idx_it.Next())
    {
        child_id = idx_it.GetPageId();
        if (child_id == MaxPageId)
        {
            break;
        }
        results[result_size++] = child_id;
    }
    result_size_dst = result_size;
    return KvError::NoError;
}

KvError IndexPageManager::SeekIndex(MappingSnapshot *mapping,
                                    PageId page_id,
                                    std::string_view key,
                                    PageId &result)
{
    auto [node, err] = FindPage(mapping, page_id);
    CHECK_KV_ERR(err);
    IndexPageIter idx_it{node, Options()};
    idx_it.Seek(key);
    PageId child_id = idx_it.GetPageId();
    if (node->IsPointingToLeaf())
    {
        result = child_id;
        return KvError::NoError;
    }
    else
    {
        return SeekIndex(mapping, child_id, key, result);
    }
}

const KvOptions *IndexPageManager::Options() const
{
    return io_manager_->options_;
}

AsyncIoManager *IndexPageManager::IoMgr() const
{
    return io_manager_;
}
}  // namespace eloqstore

File: rustdev/src/index/index_page_manager.rs
//! Index page manager implementation
//! Following the C++ index_page_manager.h/cpp

use std::collections::HashMap;
use std::sync::{Arc, RwLock, Mutex};
use std::ptr::NonNull;

use crate::config::KvOptions;
use crate::codec::Comparator;
use crate::error::{Error, KvError};
use crate::types::{PageId, TableIdent, MAX_PAGE_ID};
use crate::page::MappingSnapshot;
use crate::storage::AsyncFileManager;
use crate::Result;

use super::index_page::{MemIndexPage, IndexPageIter};
use super::root_meta::{RootMeta, CowRootMeta};

/// Index page manager for managing in-memory index pages
#[derive(Debug)]
pub struct IndexPageManager {
    /// KV options
    options: Arc<KvOptions>,
    /// File manager for async I/O
    io_manager: Arc<AsyncFileManager>,
    /// Root metadata for each table
    roots: RwLock<HashMap<TableIdent, RootMeta>>,
    /// Free list of index pages
    free_list: Mutex<Vec<Box<MemIndexPage>>>,
    /// Active list head (most recently used)
    active_head: Mutex<Option<NonNull<MemIndexPage>>>,
    /// Active list tail (least recently used)
    active_tail: Mutex<Option<NonNull<MemIndexPage>>>,
    /// Total number of pages in memory
    page_count: Mutex<usize>,
    /// Maximum pages in memory
    max_pages: usize,
}

// Safe because we carefully manage the raw pointers
unsafe impl Send for IndexPageManager {}
unsafe impl Sync for IndexPageManager {}

impl IndexPageManager {
    /// Create a new index page manager
    pub fn new(io_manager: Arc<AsyncFileManager>, options: Arc<KvOptions>) -> Self {
        let max_pages = options.max_index_pages();

        Self {
            options,
            io_manager,
            roots: RwLock::new(HashMap::new()),
            free_list: Mutex::new(Vec::new()),
            active_head: Mutex::new(None),
            active_tail: Mutex::new(None),
            page_count: Mutex::new(0),
            max_pages,
        }
    }

    /// Get comparator
    pub fn get_comparator(&self) -> Arc<dyn Comparator> {
        self.options.comparator()
    }

    /// Allocate a new index page from the pool
    pub fn alloc_index_page(&self) -> Box<MemIndexPage> {
        // Try to get from free list first
        {
            let mut free_list = self.free_list.lock().unwrap();
            if let Some(page) = free_list.pop() {
                return page;
            }
        }

        // Allocate new page
        Box::new(MemIndexPage::new(true))
    }

    /// Free an index page back to the pool
    pub fn free_index_page(&self, mut page: Box<MemIndexPage>) {
        // Reset the page
        page.unpin();
        page.deque();

        // Add to free list
        let mut free_list = self.free_list.lock().unwrap();
        free_list.push(page);
    }

    /// Enqueue an index page into the active list for cache replacement
    pub fn enqueue_index_page(&self, page: &mut MemIndexPage) {
        let mut head = self.active_head.lock().unwrap();
        let mut tail = self.active_tail.lock().unwrap();

        let page_ptr = NonNull::new(page as *mut _).unwrap();

        if head.is_none() {
            // First page in the list
            *head = Some(page_ptr);
            *tail = Some(page_ptr);
        } else {
            // Add to head (MRU position)
            unsafe {
                let head_page = head.unwrap().as_mut();
                page.enque_next(head_page);
            }
            *head = Some(page_ptr);
        }

        let mut page_count = self.page_count.lock().unwrap();
        *page_count += 1;
    }

    /// Find root metadata for a table
    /// Returns a simplified view since RootMeta can't be cloned
    pub fn find_root(&self, table_ident: &TableIdent) -> Result<RootMeta> {
        let roots = self.roots.read().unwrap();

        if let Some(root) = roots.get(table_ident) {
            // Create a simplified RootMeta copy
            // Note: This is a workaround - in production we'd need proper Arc wrapping
            let meta = RootMeta {
                root_id: root.root_id,
                ttl_root_id: root.ttl_root_id,
                mapper: None,  // Can't clone the mapper easily
                mapping_snapshots: RwLock::new(Default::default()),
                manifest_size: root.manifest_size,
                next_expire_ts: root.next_expire_ts,
                ref_cnt: Mutex::new(0),
                locked: Mutex::new(false),
                waiting: Default::default(),
            };
            return Ok(meta);
        }

        Err(Error::from(KvError::NotFound))
    }

    /// Create a COW root for write transaction
    pub fn make_cow_root(&self, table_ident: &TableIdent) -> Result<CowRootMeta> {
        let roots = self.roots.read().unwrap();

        if let Some(root) = roots.get(table_ident) {
            // Lock the root for exclusive access
            // Create a COW copy of the mapper
            let cow_meta = CowRootMeta {
                root_id: root.root_id,
                ttl_root_id: root.ttl_root_id,
                mapper: root.mapper.as_ref().map(|m| Box::new((**m).clone())),
                manifest_size: root.manifest_size,
                old_mapping: None,
                next_expire_ts: root.next_expire_ts,
            };

            return Ok(cow_meta);
        }

        // Create new COW root for new table
        Ok(CowRootMeta::default())
    }

    /// Update root metadata after commit
    pub fn update_root(&self, table_ident: &TableIdent, new_meta: CowRootMeta) {
        let mut roots = self.roots.write().unwrap();

        let root = roots.entry(table_ident.clone()).or_insert_with(RootMeta::default);

        root.root_id = new_meta.root_id;
        root.ttl_root_id = new_meta.ttl_root_id;
        root.mapper = new_meta.mapper;
        root.manifest_size = new_meta.manifest_size;
        root.next_expire_ts = new_meta.next_expire_ts;
    }

    /// Export all root metadata for manifest
    pub fn export_roots(&self) -> HashMap<TableIdent, CowRootMeta> {
        let roots = self.roots.read().unwrap();
        roots.iter()
            .map(|(table_id, root)| {
                let cow_meta = CowRootMeta {
                    root_id: root.root_id,
                    ttl_root_id: root.ttl_root_id,
                    mapper: root.mapper.clone(),
                    manifest_size: root.manifest_size,
                    next_expire_ts: root.next_expire_ts,
                    old_mapping: None, // Not needed for checkpoint
                };
                (table_id.clone(), cow_meta)
            })
            .collect()
    }

    /// Restore root metadata from manifest
    pub fn restore_roots(&self, roots_data: HashMap<TableIdent, CowRootMeta>) {
        let mut roots = self.roots.write().unwrap();
        for (table_id, cow_meta) in roots_data {
            let root = roots.entry(table_id).or_insert_with(RootMeta::default);
            root.root_id = cow_meta.root_id;
            root.ttl_root_id = cow_meta.ttl_root_id;
            root.mapper = cow_meta.mapper;
            root.manifest_size = cow_meta.manifest_size;
            root.next_expire_ts = cow_meta.next_expire_ts;
        }
    }

    /// Find a page by ID, loading from disk if necessary
    pub async fn find_page(
        &self,
        mapping: &MappingSnapshot,
        page_id: PageId,
    ) -> Result<Box<MemIndexPage>> {
        // Check if page is already in memory (swizzled pointer)
        // For now, always load from disk

        // Map logical page to file page
        let file_page_id = mapping.to_file_page(page_id)?;

        // Load page from disk
        let page_data = self.io_manager
            .read_page(file_page_id.file_id() as u64, file_page_id.page_offset())
            .await?;

        let mut index_page = self.alloc_index_page();
        index_page.set_page_id(page_id);
        index_page.set_file_page_id(file_page_id.into());
        // TODO: Copy page_data into index_page

        Ok(index_page)
    }

    /// Free a mapping snapshot
    pub fn free_mapping_snapshot(&self, mapping: *const MappingSnapshot) {
        // Find which root this mapping belongs to and remove it
        let roots = self.roots.read().unwrap();
        for (_, root) in roots.iter() {
            root.remove_mapping_snapshot(mapping);
        }
    }

    /// Unswizzle a page (convert swizzled pointer back to page ID)
    pub fn unswizzling(&self, page: &mut MemIndexPage) {
        // Convert in-memory pointer references back to page IDs
        // This is called before evicting a page
        // TODO: Implement swizzling/unswizzling logic
    }

    /// Finish I/O for a page
    pub fn finish_io(&self, mapping: &MappingSnapshot, idx_page: &mut MemIndexPage) {
        // Wake up any waiters on this page
        // TODO: Implement waiting zone notification
    }

    /// Seek in the index tree to find the data page containing a key
    pub async fn seek_index(
        &self,
        mapping: &MappingSnapshot,
        page_id: PageId,
        key: &[u8],
    ) -> Result<PageId> {
        if page_id == MAX_PAGE_ID {
            return Ok(MAX_PAGE_ID);
        }

        let mut current_page_id = page_id;

        loop {
            // Load the index page
            let index_page = self.find_page(mapping, current_page_id).await?;

            // Create iterator for the page
            let mut iter = IndexPageIter::new(&index_page, &self.options);

            // Seek to the key
            iter.seek(key);

            // Get the page ID that might contain the key
            let next_page_id = iter.get_page_id();

            // If this index page points to leaf (data pages), we're done
            if index_page.is_pointing_to_leaf() {
                return Ok(next_page_id);
            }

            // Otherwise, continue traversing down the tree
            current_page_id = next_page_id;
        }
    }

    /// Get KV options
    pub fn options(&self) -> &Arc<KvOptions> {
        &self.options
    }

    /// Get I/O manager
    pub fn io_manager(&self) -> &Arc<AsyncFileManager> {
        &self.io_manager
    }

    /// Evict root if empty
    pub fn evict_root_if_empty(&self, table_id: &TableIdent) {
        let mut roots = self.roots.write().unwrap();

        if let Some(root) = roots.get(table_id) {
            // Check if root can be evicted:
            // 1. Root page has been evicted
            // 2. No mapping snapshots active
            // 3. All pages of the tree have been evicted
            if root.root_id == MAX_PAGE_ID && !root.has_active_snapshots() && !root.is_pinned() {
                roots.remove(table_id);
            }
        }
    }

    /// Check if memory is full
    fn is_full(&self) -> bool {
        let count = self.page_count.lock().unwrap();
        *count >= self.max_pages
    }

    /// Evict a page from memory
    fn evict(&self) -> bool {
        // Find the LRU page that is not pinned
        let tail = self.active_tail.lock().unwrap();

        if let Some(tail_ptr) = *tail {
            unsafe {
                let page = tail_ptr.as_ref();
                if !page.is_pinned() {
                    // Unswizzle and evict
                    // TODO: Implement eviction
                    return true;
                }
            }
        }

        false
    }

    /// Recycle a page
    fn recycle_page(&self, page: &mut MemIndexPage) -> bool {
        if page.is_pinned() {
            return false;
        }

        // Remove from active list
        page.deque();

        // Add to free list
        self.free_index_page(Box::new(MemIndexPage::new(false)));

        let mut page_count = self.page_count.lock().unwrap();
        *page_count -= 1;

        true
    }
}